src/python/pants/core/util_rules/wrap_source.py

```diff
@@ -0,0 +1,153 @@
+from pants.engine.internals.selectors import Get
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import Rule, collect_rules, rule, rule_helper
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    GeneratedSources,
+    GenerateSourcesRequest,
+    MultipleSourcesField,
+    SourcesField,
+    SpecialCasedDependencies,
+    StringSequenceField,
+    Target,
+    Targets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.strutil import softwrap
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class WrapSource:
+    rules: tuple[Union[Rule, UnionRule], ...]
+    target_types: tuple[type[Target], ...]
+
+
+class ActivateWrapSourceTargetFieldBase(MultipleSourcesField):
+    # We solely register so that codegen can match a fieldset.
+    # One unique subclass must be defined per target type.
+    alias = "_sources"
+    uses_source_roots = False
+    expected_num_files = 0
+
+
+class WrapSourceInputsField(SpecialCasedDependencies):
+    alias = "inputs"
+    required = True
+    help = "The input targets that are to be made available by this target."
+
+
+class WrapSourceOutputsField(StringSequenceField):
+    alias = "outputs"
+    required = False
+    help = softwrap(
+        "The output files that are made available in the new context by this target. If not "
+        "specified, the target will capture all files with the expected extensions for this "
+        "source format: see the help for the target for the specific extensions. If no extensions "
+        "are specified and this value is not specified, all input files will be returned."
+    )
+
+
+@rule_helper
+async def _wrap_source(wrapper: GenerateSourcesRequest) -> GeneratedSources:
+    request = wrapper.protocol_target
+    default_extensions = {i for i in (wrapper.output.expected_file_extensions or ()) if i}
+
+    inputs = await Get(
+        Targets,
+        UnparsedAddressInputs,
+        request.get(WrapSourceInputsField).to_unparsed_address_inputs(),
+    )
+
+    sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[tgt.get(SourcesField) for tgt in inputs],
+            for_sources_types=(SourcesField, FileSourceField),
+            enable_codegen=True,
+        ),
+    )
+
+    outputs_value: Iterable[str] | None = request.get(WrapSourceOutputsField).value
+    if outputs_value:
+        pass
+    elif default_extensions:
+        outputs_value = [i for i in sources.files if any(i.endswith(j) for j in default_extensions)]
+    else:
+        outputs_value = sources.files
+
+    # I'm sure there's a better way to do this, but this works for now.
+    filter_digest = await Get(
+        ProcessResult,
+        Process(
+            argv=("/usr/bin/true",),
+            description="Filter digest",
+            input_digest=sources.snapshot.digest,
+            output_files=outputs_value,
+        ),
+    )
```

Comment:
I'm doing this to selectively filter the `output_files` out of the `input_digest`; I could not find a primitive that does that. Was I missing one?

Reply:
`await Get(Digest, DigestSubset(digest, PathGlobs(...))`